knife/lib/chef/knife/client_create.rb

```diff
@@ -81,6 +81,21 @@ def run
           client.public_key File.read(File.expand_path(config[:public_key]))
         end
 
+        # Check the file before creating the client so the api is more transactional.
+        if config[:file]
+          file = config[:file]
+
+          unless File.writable?(File.dirname(file))
+            ui.fatal "Dir #{File.dirname(file)} is not writable. Check permissions."
+            exit 1
+          end
+
+          unless File.writable?(file)
```

Comment:
If the file does not exist, the File.writable? will return false: 

![capture_036](https://user-images.githubusercontent.com/1130204/120680301-2321e980-c468-11eb-8d11-9f81b8fcdd04.png)

Reply:
yes, if the file is writable then only it will return true else false. You can also go through documentation [here](https://ruby-doc.org/core-2.5.0/File.html). Now as per your example:
1. ` File.writable?("does-not-exist")` it has checked if the given file is writable or not and returns `false` as it doesn't exist.
2. `File.write("does-not-exist", "blah")` it returns 4. Here it will create a new file "does-not-exist" and add a content "blah" can returns length of the content.
3. ` File.writable?("does-not-exist")` now after 2nd point where the file is not created now ` File.writable?` is return `true`.

Reply:
But in the case where the directory for `/tmp/test/afile.out`  does not exist, we will repor tan error that "the file '/tmp/test/' is not writeable. Check permissions".  This is misleading, and we could more actionable feedback than "check permissions".  It would be better to specifically verify that the directory exists and report that error independently of permissions-related errors.

Reply:
@marcparadise Got your point. I Will update the condition accordingly.